components/eamxx/src/share/io/scorpio_output.cpp

```diff
@@ -662,23 +678,37 @@ setup_output_file(const std::string& filename,
   set_degrees_of_freedom(filename);
 }
 /* ---------------------------------------------------------- */
+// This routine will evaluate the diagnostics stored in this
+// output instance.
+void AtmosphereOutput::compute_diagnostic(const std::string& name)
+{
+  auto skip_diag = m_diag_computed[name];
+  if (skip_diag) {
+    // Diagnostic already computed, just return
+    return;
+  }
+  const auto& diag = m_diagnostics.at(name);
+  // Check if the diagnostics has any dependencies, if so, evaluate
+  // them as well.  Needed if a diagnostic relies on another
+  // diagnostic.
+  for (const auto& dep : m_diag_depends_on_diags.at(name)) {
+    compute_diagnostic(dep);
+  }
+  diag->compute_diagnostic();
+  m_diag_computed[name] = true;
+}
+/* ---------------------------------------------------------- */
 // General get_field routine for output.
 // This routine will first check if a field is in the local field
 // manager.  If not it will next check to see if it is in the list
 // of available diagnostics.  If neither of these two options it
 // will throw an error.
-Field AtmosphereOutput::get_field(const std::string& name, const bool eval_diagnostic) const
+Field AtmosphereOutput::get_field(const std::string& name, const std::shared_ptr<const fm_type>& field_mgr) const
 {
-  if (m_field_mgr->has_field(name)) {
-    return m_field_mgr->get_field(name);
+  if (field_mgr->has_field(name)) {
+    return field_mgr->get_field(name);
   } else if (m_diagnostics.find(name) != m_diagnostics.end()) {
```

Comment:
Since we compute diags on the sim grid, we should be careful returning a diag here, in case field_mgr is not the sim FM. I think we should only check the diags if we're asking for a field from the sim FM, so we should do something ilke

```
} else {
   EKAT_REQUIRE_MSG (
     m_diagnostics.find(name) != m_diagnostics.end() && field_mgr==m_sim_field_mgr,
     "Field " + name + " not found in output field manager or diagnostics list");
```
This should protect us in case we accidentally ask a field that is not in the io FM, and we end up returning the diagnostic. It might be a corner case, but with some fields names changing (with diags), it might happen we don't register (or register with wrong name) the field in the remapper, leading to the field not being in the IO fm...